src/beast/beast/crypto/impl/sha2_context.h

```diff
@@ -80,41 +95,192 @@ struct sha512_context
 #define BEAST_SHA512_F2(x) (BEAST_SHA2_ROTR(x, 14) ^ BEAST_SHA2_ROTR(x, 18) ^ BEAST_SHA2_ROTR(x, 41))
 #define BEAST_SHA512_F3(x) (BEAST_SHA2_ROTR(x,  1) ^ BEAST_SHA2_ROTR(x,  8) ^ BEAST_SHA2_SHFR(x,  7))
 #define BEAST_SHA512_F4(x) (BEAST_SHA2_ROTR(x, 19) ^ BEAST_SHA2_ROTR(x, 61) ^ BEAST_SHA2_SHFR(x,  6))
-#define BEAST_SHA2_UNPACK32(x, str)             \
+#define BEAST_SHA2_PACK32(str, x)               \
 {                                               \
-    *((str) + 3) = (std::uint8_t) ((x)      );         \
-    *((str) + 2) = (std::uint8_t) ((x) >>  8);         \
-    *((str) + 1) = (std::uint8_t) ((x) >> 16);         \
-    *((str) + 0) = (std::uint8_t) ((x) >> 24);         \
+    *(x) =                                      \
+        ((std::uint32_t) *((str) + 3)      )    \
+      | ((std::uint32_t) *((str) + 2) <<  8)    \
+      | ((std::uint32_t) *((str) + 1) << 16)    \
+      | ((std::uint32_t) *((str) + 0) << 24);   \
 }
-#define BEAST_SHA2_UNPACK64(x, str)             \
+#define BEAST_SHA2_UNPACK32(x, str)             \
 {                                               \
-    *((str) + 7) = (std::uint8_t) ((x)      );         \
-    *((str) + 6) = (std::uint8_t) ((x) >>  8);         \
-    *((str) + 5) = (std::uint8_t) ((x) >> 16);         \
-    *((str) + 4) = (std::uint8_t) ((x) >> 24);         \
-    *((str) + 3) = (std::uint8_t) ((x) >> 32);         \
-    *((str) + 2) = (std::uint8_t) ((x) >> 40);         \
-    *((str) + 1) = (std::uint8_t) ((x) >> 48);         \
-    *((str) + 0) = (std::uint8_t) ((x) >> 56);         \
+    *((str) + 3) = (std::uint8_t) ((x)      );  \
+    *((str) + 2) = (std::uint8_t) ((x) >>  8);  \
+    *((str) + 1) = (std::uint8_t) ((x) >> 16);  \
+    *((str) + 0) = (std::uint8_t) ((x) >> 24);  \
 }
 #define BEAST_SHA2_PACK64(str, x)               \
 {                                               \
-    *(x) =   ((std::uint64_t) *((str) + 7)      )      \
-           | ((std::uint64_t) *((str) + 6) <<  8)      \
-           | ((std::uint64_t) *((str) + 5) << 16)      \
-           | ((std::uint64_t) *((str) + 4) << 24)      \
-           | ((std::uint64_t) *((str) + 3) << 32)      \
-           | ((std::uint64_t) *((str) + 2) << 40)      \
-           | ((std::uint64_t) *((str) + 1) << 48)      \
-           | ((std::uint64_t) *((str) + 0) << 56);     \
+    *(x) =                                      \
+          ((std::uint64_t) *((str) + 7)      )  \
+        | ((std::uint64_t) *((str) + 6) <<  8)  \
+        | ((std::uint64_t) *((str) + 5) << 16)  \
+        | ((std::uint64_t) *((str) + 4) << 24)  \
+        | ((std::uint64_t) *((str) + 3) << 32)  \
+        | ((std::uint64_t) *((str) + 2) << 40)  \
+        | ((std::uint64_t) *((str) + 1) << 48)  \
+        | ((std::uint64_t) *((str) + 0) << 56); \
+}
+#define BEAST_SHA2_UNPACK64(x, str)             \
+{                                               \
+    *((str) + 7) = (std::uint8_t) ((x)      );  \
```

Comment:
Don't we have all sorts of code to do this?  And is that cast really needed, how could the compiler get it wrong?